swaybar/tray/tray.c

```diff
@@ -90,9 +91,17 @@ void destroy_tray(struct swaybar_tray *tray) {
 }
 
 void tray_in(int fd, short mask, void *data) {
-	sd_bus *bus = data;
+	struct swaybar *bar = data;
 	int ret;
-	while ((ret = sd_bus_process(bus, NULL)) > 0) {
+
+	if (mask & (POLLHUP | POLLERR)) {
+        sway_log(SWAY_ERROR, "D-Bus connection closed unexpectedly");
+		sd_bus_unref(bar->tray->bus);
```

Comment:
It sounds like this `sd_bus_unref()` call is something which should be done by `bar_teardown()` or similar, instead of being the responsibility of all spots we set `running = false`.